includes/class-wc-stripe-api.php

```diff
@@ -250,6 +278,80 @@ public static function retrieve( $api ) {
 		return json_decode( $response['body'] );
 	}
 
+	/**
+	 * Checks if the Stripe API for the current mode (i.e. test or live) is rate limited.
+	 *
+	 * @return bool True if the Stripe API is rate limited, false otherwise.
+	 */
+	public static function is_stripe_api_rate_limited() {
+		$rate_limit_option_key = WC_Stripe_Mode::is_test() ? self::TEST_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY : self::LIVE_MODE_STRIPE_API_RATE_LIMIT_OPTION_KEY;
+
+		$rate_limit_expiration = get_option( $rate_limit_option_key );
+		if ( ! $rate_limit_expiration ) {
+			return false;
+		}
+
+		$now = time();
+		if ( $now > $rate_limit_expiration ) {
+			delete_option( $rate_limit_option_key );
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Helper function to check error responses from Stripe and ensure we prevent unnecessary API calls,
+	 * primarily in cases where we have been rate limited or we don't have valid keys..
+	 *
+	 * @param array|WP_Error $response The response from the Stripe API.
+	 * @return void
+	 */
+	protected static function check_stripe_api_error_response( $response ) {
+		// If we don't have an array for $response, return early, as we won't have an HTTP status code.
+		if ( ! is_array( $response ) ) {
+			return;
+		}
+
+		// We specifically want to check $response['response']['code']. If it's not present, return early.
+		if ( ! isset( $response['response'] ) || ! is_array( $response['response'] ) || ! isset( $response['response']['code'] ) ) {
```

Comment:
Nit, only for readability: I think it's sufficient to just check `! isset( $response['response']['code'] )` here?

```
❯ php -a
Interactive shell

php > $a = [];
php > echo $a['b']['c'];
PHP Warning:  Undefined array key "b" in php shell code on line 1

Warning: Undefined array key "b" in php shell code on line 1
PHP Warning:  Trying to access array offset on null in php shell code on line 1

Warning: Trying to access array offset on null in php shell code on line 1
php > var_dump( isset( $a['b']['c'] ) );
bool(false)
```

Reply:
We could also split it into multiple lines, but the issue is if `$response['response']` is an object, even if we use `isset()`.
```
$ php -a
Interactive shell

php > $b = (object) [ 'c' => 'c' ]; $a = [ 'b' => $b ];
php > echo isset( $a['b']['c'] );
PHP Warning:  Uncaught Error: Cannot use object of type stdClass as array in php shell code:1
Stack trace:
#0 {main}
  thrown in php shell code on line 1

Warning: Uncaught Error: Cannot use object of type stdClass as array in php shell code:1
Stack trace:
#0 {main}
  thrown in php shell code on line 1
```

Reply:
I've split this into a multiline OR condition to help with the readability in 4e638b87.